### ISSUE TITLE
Create validation error on failed Transform

### ISSUE THREAD

**Issue by ShadowDancer** [FluentValidation/FluentValidation#1785]
### Is your feature request related to a problem? Please describe.

I want to return an validation result when transform fails.
Right now exception bubbles up into `.Validate` method, which kind of breaks point of having such validation framework in the first place.

```cs
validator.Validate(objectWithFailingTransform)
```
Will throw an exception. What is a point of validation if I have to catch exception myself?

### Describe the solution you'd like

There should be some way to report an error. For instance magine you have 

```cs
class JsonWrapper { string JsonProp {get; set; } }
class JsonModel { int Value {get; set;} }

class JsonWrapperValidator : AbstractValidator<JsonWrapper> { 
   public JsonWrapperValidator (){ 
       Transform(p=> p.JsonProp, jsonString => _serializer.Deserialize<JsonModel >(jsonString))
                .WithDefaultError("{PropertyName} must be a valid JSON")
                .RuleFor(p => p.Value).InclusiveBetween(1, 5);
   }
}
```
`new JsonWrapperValidator().Validate` would return ValidationResult with error that JsonProp must ba a valid JSON.
Currently you have to implement such mechanism yourself, which is really weird.

Also I imagine it would be useful to be able to customize message. I think most flexible way would be to be able to throw `TransformException` with path and message. For instance:
```cs
class JsonWrapperValidator : AbstractValidator<JsonWrapper> { 
   public JsonWrapperValidator (){ 
       Transform(p=> p.JsonProp, jsonString => { 
                if(jsonString == null) throw new TransformException("{PropertyName} cannot be null");
                if(jsonString.Length < 256) throw new TransformException("{PropertyName} can be at most 256 characters long");
                try {
                      _serializer.Deserialize<JsonModel >(jsonString)})
                }
                catch(JsonSerializerException ex)
                {
                        throw new TransformException("{PropertyName} must be a valid json", ex);
                })
                .WithDefaultMessage("Validation of {PropertyName} failed.") // this probably should be a default message
                .RuleFor(p => p.Value).InclusiveBetween(1, 5);
   }
}
```

### Additional Context

https://github.com/FluentValidation/FluentValidation/issues/1631

I would like to contribute such feature.

**Comment by ShadowDancer** [comment]
I think it'd be best to do it without breaking changes in api (i.e. throwing exception) or adding overloads.

**Comment by ShadowDancer** [comment]
Also it could be useful to plan one step further and allow pretransform validation. I.e. 
```cs
class JsonWrapperValidator : AbstractValidator<JsonWrapper> { 
   public JsonWrapperValidator (){ 
       Transform(p=> p.JsonProp, jsonString => { 
                if(jsonString == null) throw new TransformException("{PropertyName} cannot be null");
                if(jsonString.Length < 256) throw new TransformException("{PropertyName} can be at most 256 characters long");
                try {
                      _serializer.Deserialize<JsonModel >(jsonString)})
                }
                catch(JsonSerializerException ex)
                {
                        throw new TransformException("{PropertyName} must be a valid json", ex);
                })
                .WithDefaultMessage("Validation of {PropertyName} failed.") 
                .PreTransformRules().NotNull().Length(1, 256)
                .RuleFor(p => p.Value).InclusiveBetween(1, 5);
   }
}

```
You may ask why I need to validate pre transform rules, and the answer is:

I have jsons stored as text in database. I need to check that maximum length is acceptable, before letting request through. But even if I check length, I still want it to be valid json, and maybe apply validators to deserialized object.

**Comment by JeremySkinner** [comment]
Hi, thanks for the suggestion.

Generally I feel that complex transformations aren't something that should be done by FluentValidation, so I'm not sure this feature would be a good fit for the library. FluentValidation's transformations are intended for simple use cases, eg where you want to treat a string property as an integer solely for the purpose of validation (and not reuse the transformed value afterwards). 

More complex transformation logic, or transformations where you need to use the transformed value outside of the validator really aren't the job of the validator - this is a separate process and you should handle it externally, storing the transformed value on a separate property which can then be validated separately. Although data transformations and validation do overlap, they really are separate processes and should be treated as such. 

In your specific example, even if we were to implement this the usefulness would be pretty limited, as the transformed value would be lost after validation has finished, and you'd need to deserialize it a second time to use the value afterwards (as FluentValidation only reads properties, it doesn't set them). This would be incredibly inefficient. It would be much better to do this as several distinct steps:
- Have a validator for performing the string validation on the raw data (this would be the equivalent of your "pre-transform" rules)
- Invoke this validator
- If it succeeds, attempt to deserialize the string to the type - handle this check manually
- If it succeeds, pass the deserialized data to a second validator to handle any validation of the deserialized data.   

This might not sound as "clean", but it's actually much more efficient as you only need to deserialize once, and it also keeps the individual validators focused on a single task. 

Because of the above reasons I don't think this is something we'd include in the core project, but I appreciate the suggestion. 

Hope that makes sense.

**Comment by ShadowDancer** [comment]
```
FluentValidation's transformations are intended for simple use cases, eg where you want to treat a string property as an integer solely for the purpose of validation (and not reuse the transformed value afterwards)
```
It feels like user trap where someone invests time into building infrastructure around validating transformed value, and then discovers that he has to rewrite all of his validators to using source type, because library cannot handle transformation error. Even in most simple case like parsing int from string there are strings which are not integers and you have to manually write methods like `InclusiveBetweenIntFromStr`.

What's the point of having half-baked Transform and AspNetCore integration in the library then?

```
In your specific example, even if we were to implement this the usefulness would be pretty limited, as the transformed value would be lost after validation has finished
```
Which I don't care about, because it goes to database as string anyway. But even if it didn't deserializing twice is cheap. Maintaining duplicate dto classes with additional property and maintaining code mapping between the two is not cheap.

Maybe I've overengineered this case, but I think that there should be SOME way to handle errors for Transform gracefully.

**Comment by JeremySkinner** [comment]
> It feels like user trap where someone invests time into building infrastructure around validating transformed value

How is this a “trap”? The documentation is quite clear about what the feature does. If it isn’t suitable for your use case then you can just take a different approach, whether that’s by using a custom rule, or a two step process or even not using FluentValidation, depending on the individual circumstances and situation. 

> What's the point of having half-baked Transform and AspNetCore integration in the library then?

“Half baked”. Ok cool. 

FluentValidation was built in a way that provides methods for handling simple use cases of the box, and then you can drop down to a lower level for scenarios that are more complex. This applies to many areas of the library including Transformations. This isn’t “half baked” but is an intentional design decision. For transformations specifically, the current feature is suitable for many simple uses cases where al you need is a simple type check, and a fallback to null if the transform fails (which cpile be caught with a NotNull rule).  If your situation needs something more complex then a surrogate property with explicit transformation logic is a better approach. That doesn’t have to be done completely externally (although that’s usually my preference) 

> AspNetCore integration

I’m not sure how this is relevant as we weren’t discussing asp.net, but as you mentioned it I’ll point out a couple of things. Firstly, the asp.net integration is easily the most complex piece of the library and has received a huge amount of work (hundreds of man hours) over many years to get asp.net’s validation infrastructure to play nicely with FluentValidation, and is constantly updated to work with the updates made to asp.net core. I’m not sure which part of this you’re specifically referring to as half-baked but are some things that simply can’t be done because of the limitations of asp.net’s validation api, but despite this it continues to receive development and maintenance. If you want to open an issue to report something that isn’t working correctly in the asp.net integration then please  feel free, but dismissing this work as “half baked” is completely uncalled for.

> Maintaining duplicate dto classes with additional property and maintaining code mapping between the two is not cheap.

Personally I disagree, but of course you’re welcome to your opinion and you don’t have to do it that way if you don’t want to…you could use a single class and a single validator with surrogate properties, or keep it all in a single validator with a custom rule, or some other way that I’ve not thought of.

I’ll finish with this: Honestly your response came across as incredibly rude. I maintain this project for free in my spare time, and I took the effort to engage with you and explain my reasoning in good faith. There’s really no need to be so hostile.

**Comment by ShadowDancer** [comment]
Sorry if my response come rude, I didn't mean that.

By "half-baked' I meant that I couldn't use your proposed solution together with AspNetCore validation, because everything happens in filter. I didn't mean to be offensive, I just felt that this could use some improvements in error handling area - which I wanted to contribute to, especially considering fact that you thought about it already in linked issue. Maybe I didn't understood your intentions well enough, I just thought that's some grey area of the library which could be explored.

`The documentation is quite clear about what the feature does.`
I think that documentation is clear enough about that it does, but it's not clear that it's extremely limited, and you cannot handle errors in reasonable way. I.e. even example from documentation is not really useful, as reported error will be `NumberProperty cannot be empty`, not `NumberProperty is not a number`.
I kinda think about if like trap because in programming you're used to creating and reusing concepts. When I look at library it feels really intuitive and natural for me to transform one value to another and then reuse existing validation logic, and handle additional case where this cannot even be transformed, but you're better of doing this as special rules which validate source type, so you have control over the process.
I think this is especially true considering existence of some advanced features related to transform, like chaining properties of transformed value to parent properties (https://dotnetfiddle.net/Widget/3qz20j).

`If your situation needs something more complex then a surrogate property with explicit transformation logic is a better approach.`
But this recommended approach does not play nice with AspNetCore integration, that's what I meant by half-baked. Again, I don't want to be offensive, just these two features don't click with each other and they could, with relatively minor addition, which I thought you already liked.

Again sorry for making you feel offended, I just wanted to highlight that these two parts of library could play together in really useful way, didn't mean to make any negative feelings.

**Comment by JeremySkinner** [comment]
> Sorry if my response come rude, I didn't mean that.

Thank you, I appreciate that. 

> even example from documentation is not really useful, as reported error will be NumberProperty cannot be empty, not NumberProperty is not a number

Yes that's a good point, you would need some way of differentiating a 'missing' null vs a 'transform failed'. In that case it might be better for the transformer to return a tuple of value/success where you'd return `(deserializedValue, true)` on a successful deserialization and `(null, false)` when failed. Although this still isn't perfect as, it may be suitable for some situations. 

> But this recommended approach does not play nice with AspNetCore integration,

I believe a surrogate property (or method) should work OK with the aspnetcore integration if you keep the deserialization logic within the surrogate rather than doing it as a 2-step process. As a quick example I imagine it'd look something like this:

```csharp

class JsonWrapper 
{ 
  public string JsonProp {get; set; } 

  public (JsonModel Value, bool Success) JsonDeserialized 
  {
    get 
    {
       try 
       {
         var deserialized = Serializer.Deserialize<JsonModel>(JsonProp); 
         // optionally could cache to save multiple deserializations ​
        ​return (deserialized, true);
      ​}
      ​catch  
      ​{
         ​return (null, false);
      ​}
   ​}
 ​}  
}
​
class MyValidator : AbstractValidator<JsonWrapper> 
{
  ​ctor() 
  ​{
      ​RuleFor(x => x.JsonProp).NotNull().Length(1, 256);
      ​RuleFor(x => x.JsonDeserialized,Success).NotEqual(false).WithMessage("Must be valid json").OverridePropertyName("JsonProp");
      RuleFor(x => x.JsonDeserialized.Value).InclusiveBetween(1, 5).When(x => x.JsonDeserialized.Success).OverridePropertyName("JsonProp");
​  }
}
```
This approach would keep the validator focused on just validation and free of any serialization logic, and by having the deserialization take place inside the surrogate property this would mean it'd work ok with ASP.NET's model binding. Alternatively if you don't like having the serialization logic inside the class itself then you could put it in somewhere else that gets invoked prior to validation (for example a custom model binder, or a FluentValidation interceptor, or even by overriding `Validate` on the validator class) and set the surrogate property in there. 

I think this approach is more flexible in the long run and more suited to complex transformations, but I agree that it needs more moving parts. 

Let me think about this a bit further; I'll reopen this for now and have a think about it tomorrow.

**Comment by JeremySkinner** [comment]
I’ve spent some more time thinking about this, but I’m going to stick with my original decision: transformations should be done prior to the validator running, and trying to build a mechanism for performing complex transformations into the library is out of scope for the project. 

Throwing an exception to indicate transformation failure also isn’t something that we’ll implement as 1) exceptions shouldn’t be used to control program flow and 2) they have a significant performance cost. 

I may actually end up deprecating transformations from FluentValidation completely, but haven’t decided on that yet. 

I’m sorry if that’s not the answer you wanted.

**Comment by ShadowDancer** [comment]
That's kind of a answer I wanted, because I'm on a clear ground now. 😊 Thanks for looking into this!
